### PR TITLE
Fix pruning missed nodes without caching db.

### LINKED ISSUE #6555: Trie exceptions during the full pruning
In 1.25.1 @kamilchodola noticed a few trie exceptions during full pruning.

### PR DESCRIPTION
**Pull request by asdacap**
Probably Fix #6555

- For reason beyond the understanding of mortal mind, trying to persist dirty cache does not work. I'm guessing some of the commitset points to node that is no longer in the cache... probably.
- In any case, this PR changes the strategy by persisting all commitset, then running memory pruning, clearing the cache. Hence, nothing need to be saved... probably. As long as it work...
- Additionally, previously persist cache run asyncronously which can take so much time that copying node begins. If pruning is not throttled, this can slows down persist cache dramatically, so much so that copying node can complete before persist cache complete, which then switch the database causing trieexception immediately. 
- This refactor full pruner code so that persist cache will run completely first before copying node.
- Turns out persist cache is slow because it keep re-persisting the same node several time. Just having a set that keep track of what persisted dramatically reduce the time it take to persist cache.
- The whole persist cache thingy is before I tried the clearing cache method which finally work reliably... so.. I already did the work.. might as well.
- Also because of the batching of the writes during copy, it is possible that during dispose, a read from outside could pop a writebatch from the concurrent queue causing it to be missed from persist. This PR also address that by extracting the logic into WriteBatcher class which is then used inside the CopyTreeVisitor.

## Changes

- Refactored full pruning code. Pruning does not start copying until after persist cache.
- Persist cache now clear cache.

#### What types of changes does your code introduce?

- [X] Bugfix (a non-breaking change that fixes an issue)

#### Requires testing

- [X] Yes
- [ ] No

#### If yes, did you write tests?

- [ ] Yes
- [X] No

#### Notes on testing

- Ran 10 consecutive prune-verify trie cycle on a 9 oct backup. Previously this reliably fail likely due to constant block processing.

### CODE REVIEW

File: src/Nethermind/Nethermind.Blockchain/FullPruning/FullPruner.cs [hunk 1]
```diff
@@ -90,82 +91,102 @@ private void OnPrune(object? sender, PruningTriggerEventArgs e)
...
-                // we mark that we are waiting for block (for thread safety)
-                else if (Interlocked.CompareExchange(ref _waitingForBlockProcessed, 1, 0) == 0)
+                else
                 {
-                    // we don't want to start pruning in the middle of block processing, lets wait for new head.
-                    _blockTree.OnUpdateMainChain += OnUpdateMainChain;
                     e.Status = PruningStatus.Starting;
+
+#pragma warning disable CS4014 // Because this call is not awaited, execution of the current method continues before the call is completed
+                    RunFullPruning(_processExitSource.Token);
```
[the remark this answers is not in the document]
Comment (asdacap) [reply src/Nethermind/Nethermind.Blockchain/FullPruning/FullPruner.cs]: Done

File: src/Nethermind/Nethermind.Blockchain/FullPruning/FullPruner.cs [hunk 2]
```diff
@@ -217,49 +236,40 @@ protected virtual void RunPruning(IPruningContext pruning, Hash256 statRoot)
...
+                        return true;
+                    });
 
-                    _blockTree.OnUpdateMainChain += CommitOnNewBLock;
-                    copyTreeVisitor.Finish();
-                }
-                else
-                {
-                    pruning.Dispose();
+                    _lastPruning = DateTime.Now;
```
Comment (benaadams) [inline src/Nethermind/Nethermind.Blockchain/FullPruning/FullPruner.cs:260]: Really should use `DateTime.UtcNow` for system type dates; .Now can do weird things on time changes (daylight savings, timezone changes if laptop)

**Review by benaadams — APPROVED** [review]

File: src/Nethermind/Nethermind.Blockchain/FullPruning/FullPruner.cs (hunk 2 above)
Comment (asdacap) [reply src/Nethermind/Nethermind.Blockchain/FullPruning/FullPruner.cs]: Done

File: src/Nethermind/Nethermind.Trie/Pruning/ReadOnlyTrieStore.cs [hunk 3]
```diff
@@ -52,6 +54,11 @@ public void Set(in ValueHash256 hash, byte[] rlp)
         {
         }
 
+        public void PersistCache(CancellationToken cancellationToken)
+        {
+            _trieStore.PersistCache(cancellationToken);
```
Comment (benaadams) [inline src/Nethermind/Nethermind.Trie/Pruning/ReadOnlyTrieStore.cs:59]: Should `ReadOnlyTrieStore` persist anything? 🤔

Comment (asdacap) [reply src/Nethermind/Nethermind.Trie/Pruning/ReadOnlyTrieStore.cs:59]: `ITrieStore` probably should not even have this. There should be a separate `PruningTrieStore` which actually have this method.

File: src/Nethermind/Nethermind.Blockchain.Test/FullPruning/FullPrunerTests.cs [hunk 4]
```diff
@@ -221,15 +221,18 @@ public async Task<bool> WaitForPruning()
...
+                    AddBlocks(1);
+                }
                 AddBlocks(1);
                 return await context.DisposeEvent.WaitOneAsync(TimeSpan.FromMilliseconds(Timeout.MaxWaitTime * 5), CancellationToken.None);
             }
 
             public TestFullPruningDb.TestPruningContext WaitForPruningStart()
             {
                 PruningTrigger.Prune += Raise.Event<EventHandler<PruningTriggerEventArgs>>();
+                Thread.Sleep(10);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Blockchain.Test/FullPruning/FullPrunerTests.cs:235]: Can we avoid Thread.Sleep in the tests? Expose some task to await in tests or something?

File: src/Nethermind/Nethermind.Blockchain/FullPruning/CopyTreeVisitor.cs [hunk 5]
```diff
@@ -62,7 +65,7 @@ public void VisitMissingNode(Hash256 nodeHash, TrieVisitContext trieVisitContext
             }
 
             // if nodes are missing then state trie is not valid and we need to stop copying it
-            _pruningContext.CancellationTokenSource.Cancel();
+            throw new TrieException($"Trie {nodeHash} missing");
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Blockchain/FullPruning/CopyTreeVisitor.cs:68]: add to method:
    [DoesNotReturn]
    [StackTraceHidden]

File: src/Nethermind/Nethermind.Blockchain/FullPruning/FullPruner.cs [hunk 6]
```diff
@@ -90,82 +91,100 @@ private void OnPrune(object? sender, PruningTriggerEventArgs e)
...
         }
 
-        private void OnUpdateMainChain(object? sender, OnUpdateMainChainArgs e)
+        private async Task WaitForMainChainChange(CancellationToken cancellationToken, Func<OnUpdateMainChainArgs, bool> handler)
+        {
+            await Wait.ForEventCondition<OnUpdateMainChainArgs>(
+                cancellationToken,
+                (h) => _blockTree.OnUpdateMainChain += h,
+                (h) => _blockTree.OnUpdateMainChain -= h,
+                (e) =>
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Blockchain/FullPruning/FullPruner.cs:109]: parenthesis not needed?

File: src/Nethermind/Nethermind.Blockchain/FullPruning/FullPruner.cs [hunk 7]
```diff
@@ -90,82 +91,100 @@ private void OnPrune(object? sender, PruningTriggerEventArgs e)
...
 
-        private void SetCurrentPruning(IPruningContext pruningContext)
+        private async Task RunFullPruning(IPruningContext pruningContext, CancellationToken cancellationToken)
         {
-            IPruningContext? oldPruning = Interlocked.Exchange(ref _currentPruning, pruningContext);
-            if (oldPruning is not null)
+            _trieStore.PersistCache(cancellationToken);
+
+            long blockToWaitFor = 0;
+            await WaitForMainChainChange(cancellationToken, (e) =>
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Blockchain/FullPruning/FullPruner.cs:152]: parenthesis not needed?

File: src/Nethermind/Nethermind.Blockchain/FullPruning/FullPruner.cs [hunk 8]
```diff
@@ -90,82 +91,100 @@ private void OnPrune(object? sender, PruningTriggerEventArgs e)
...
+            {
+                if (e.Blocks == null || e.Blocks.Count == 0) return false;
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Blockchain/FullPruning/FullPruner.cs:154]: `is null`, or even better use `?.`

File: src/Nethermind/Nethermind.Blockchain/FullPruning/FullPruner.cs [hunk 9]
```diff
@@ -90,82 +91,100 @@ private void OnPrune(object? sender, PruningTriggerEventArgs e)
...
-                    _blockTree.OnUpdateMainChain += OnUpdateMainChain;
                     e.Status = PruningStatus.Starting;
+
+                    _ = RunFullPruning(_processExitSource.Token);
                 }
             }
         }
 
-        private void OnUpdateMainChain(object? sender, OnUpdateMainChainArgs e)
+        private async Task WaitForMainChainChange(CancellationToken cancellationToken, Func<OnUpdateMainChainArgs, bool> handler)
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Blockchain/FullPruning/FullPruner.cs:103]: general convention is to put cancellation token at the last parameter (as it often has default)

File: src/Nethermind/Nethermind.Blockchain/FullPruning/CopyTreeVisitor.cs [hunk 10]
```diff
@@ -21,25 +23,26 @@ namespace Nethermind.Blockchain.FullPruning
...
         private long _persistedNodes = 0;
         private bool _finished = false;
         private readonly WriteFlags _writeFlags;
         private readonly CancellationToken _cancellationToken;
         private const int Million = 1_000_000;
+        private WriteBatcher _writeBatcher;
 
         public CopyTreeVisitor(
             IPruningContext pruningContext,
+            CancellationToken cancellationToken,
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Blockchain/FullPruning/CopyTreeVisitor.cs:37]: general convention is to put cancellation token at the last parameter (as it often has default)

File: src/Nethermind/Nethermind.Core/Utils/WriteBatcher.cs [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+using System;
+using System.Collections.Concurrent;
+using System.Threading;
+
+namespace Nethermind.Core.Utils;
+
+/// <summary>
+/// Batches writes into a set of concurrent batches. For cases where throughput matter, but not atomicity.
+/// </summary>
+public class WriteBatcher : IWriteBatch
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Core/Utils/WriteBatcher.cs:13]: ConcurrentWriteBatcher?

File: src/Nethermind/Nethermind.Core/Utils/WriteBatcher.cs [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+    {
+        if (_disposing) throw new InvalidOperationException("Trying to set while disposing");
+        if (!_batches.TryDequeue(out IWriteBatch? currentBatch))
+        {
+            currentBatch = _underlyingDb.StartWriteBatch();
+        }
+
+        currentBatch.PutSpan(key, value, flags);
+        long val = Interlocked.Increment(ref _counter);
+        if (val % 10000 == 0)
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Core/Utils/WriteBatcher.cs:44]: magic number -> const - or maybe a parameter?

File: src/Nethermind/Nethermind.Core/Utils/WriteBatcher.cs [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+        {
+            currentBatch.Dispose();
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Core/Utils/WriteBatcher.cs:46]: potential concurrency issue -> Can you be disposing while other thread calls `PutSpan`? Do we need `_counter < limit` somewhere before `PutSpan`?

Maybe not as ConcurrentQueue would handle it?

File: src/Nethermind/Nethermind.Core/Utils/WriteBatcher.cs [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+        }
+        else
+        {
+            _batches.Enqueue(currentBatch);
+        }
+    }
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Core/Utils/WriteBatcher.cs:72]: Basically same code as PutSpan, not sure if we can make it sharable?

File: src/Nethermind/Nethermind.Trie/Pruning/ReadOnlyTrieStore.cs (hunk 3 above)
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.Trie/Pruning/ReadOnlyTrieStore.cs:59]: Maybe we should make `IPersistableTrieStore`, put it only there?

File: src/Nethermind/Nethermind.Blockchain/FullPruning/FullPruner.cs (hunk 7 above)
Comment (asdacap) [reply src/Nethermind/Nethermind.Blockchain/FullPruning/FullPruner.cs:152]: It has some log.

File: src/Nethermind/Nethermind.Core/Utils/WriteBatcher.cs (hunk 13 above)
Comment (asdacap) [reply src/Nethermind/Nethermind.Core/Utils/WriteBatcher.cs:46]: Yes, that is why the concurrent queue is there.

The return of `Interlocked.Increment` should be atomic so it should persist it eventually. Its a bit random though.

Comment (asdacap) [reply src/Nethermind/Nethermind.Core/Utils/WriteBatcher.cs:46]: If not, dispose should dequeue everything.

File: src/Nethermind/Nethermind.Trie/Pruning/ReadOnlyTrieStore.cs (hunk 3 above)
Comment (asdacap) [reply src/Nethermind/Nethermind.Trie/Pruning/ReadOnlyTrieStore.cs]: Done, put into IPruningTrieStore.

File: src/Nethermind/Nethermind.Blockchain/FullPruning/CopyTreeVisitor.cs (hunk 5 above)
Comment (asdacap) [reply src/Nethermind/Nethermind.Blockchain/FullPruning/CopyTreeVisitor.cs:71]: Done

File: src/Nethermind/Nethermind.Blockchain/FullPruning/FullPruner.cs (hunk 8 above)
Comment (asdacap) [reply src/Nethermind/Nethermind.Blockchain/FullPruning/FullPruner.cs]: Done

File: src/Nethermind/Nethermind.Core/Utils/WriteBatcher.cs (hunk 11 above)
Comment (asdacap) [reply src/Nethermind/Nethermind.Core/Utils/WriteBatcher.cs]: Done

File: src/Nethermind/Nethermind.Core/Utils/WriteBatcher.cs (hunk 12 above)
Comment (asdacap) [reply src/Nethermind/Nethermind.Core/Utils/WriteBatcher.cs]: Added a constant.

File: src/Nethermind/Nethermind.Core/Utils/WriteBatcher.cs (hunk 14 above)
Comment (asdacap) [reply src/Nethermind/Nethermind.Core/Utils/WriteBatcher.cs]: Added rent and return batch.

File: src/Nethermind/Nethermind.Blockchain.Test/FullPruning/FullPrunerTests.cs (hunk 4 above)
Comment (asdacap) [reply src/Nethermind/Nethermind.Blockchain.Test/FullPruning/FullPrunerTests.cs]: Updated to reduce sleep. Can't find a way to eliminate them.

**Pull request merged**